packages/core/src/validate/run.ts

```diff
@@ -206,27 +225,159 @@ function* getConstructorErrors(contractDef: ContractDefinition, decodeSrc: SrcDe
   }
 }
 
-function* getOpcodeErrors(contractDef: ContractDefinition, decodeSrc: SrcDecoder): Generator<ValidationErrorOpcode> {
-  for (const fnCall of findAll('FunctionCall', contractDef, node => skipCheck('delegatecall', node))) {
+function* getOpcodeErrors(
+  contractOrFunctionDef: ContractDefinition,
+  deref: ASTDereferencer,
+  decodeSrc: SrcDecoder,
+): Generator<ValidationErrorOpcode> {
+  yield* getContractOpcodeErrors(
+    contractOrFunctionDef,
+    deref,
+    decodeSrc,
+    {
+      kind: 'delegatecall',
+      pattern: /^t_function_baredelegatecall_/,
+    },
+    false,
+    new Set<number>(),
+  );
+  yield* getContractOpcodeErrors(
+    contractOrFunctionDef,
+    deref,
+    decodeSrc,
+    {
+      kind: 'selfdestruct',
+      pattern: /^t_function_selfdestruct_/,
+    },
+    false,
+    new Set<number>(),
+  );
+}
+
+function* getContractOpcodeErrors(
+  contractDef: ContractDefinition,
+  deref: ASTDereferencer,
+  decodeSrc: SrcDecoder,
+  opcode: OpcodePattern,
+  skipInternal: boolean,
+  visitedNodeIds: Set<number>,
+): Generator<ValidationErrorOpcode> {
+  if (visitedNodeIds.has(contractDef.id)) {
+    return;
```

Comment:
Let's add a comment explaining why the early return is correct. Something like: "We return early here, but the errors from this node were emitted when the node was marked as visited.